sound/soc/sof/ipc3-topology.c

```diff
@@ -2307,12 +2308,14 @@ static int sof_ipc3_tear_down_all_pipelines(struct snd_sof_dev *sdev, bool verif
 	 * topology loading the sound card unavailable to open PCMs.
 	 */
 	list_for_each_entry(swidget, &sdev->widget_list, list) {
-		if (swidget->dynamic_pipeline_widget)
+		if (swidget->dynamic_pipeline_widget) {
+			++dyn_widgets;
 			continue;
+		}
 
-		/* Do not free widgets for static pipelines with FW ABI older than 3.19 */
+		/* Do not free widgets for static pipelines with FW older than SOF2.2 */
 		if (!verify && !swidget->dynamic_pipeline_widget &&
-		    v->abi_version < SOF_ABI_VER(3, 19, 0)) {
+		    SOF_FW_VER(v->major, v->minor, v->micro) < SOF_FW_VER(2, 2, 0)) {
```

Comment:
we don't have any other case where we test the firmware version, so this seems like an admission of failure in ABI handling? Is this saying that we have firmware newer that ABI 3.19 that doesn't work well?

I am not opposed to this, just that we need to be transparent on what happened.

Reply:
@plbossart This is not about ABI, so the old/existing usage is a bit of a misusage of the SOF_ABI_VER macro. Old firmware version did implement the interfaces to free components, but they just didn't work. So problem is not in the interface, but rather in the implementation.

This is of course a bit academic as in practise referring to ABI always refers to a SOF version as well. But it does confuse the reader, so I'd rather fix this.

Reply:
Commit subject: "ASoC: SOF: ipc3-topology: ..." or just "ASoC: SOF: ipc3: ...."

@kv2019i, it is going to be fixed in 2.3 release? Will a fixed firmware before the 2.3 work fine?

Reply:
Should this check take into account the `dyn_widgets` like you do later?

Reply:
@ujfalusi Ack on the commit subject, will change.

This code on L2318 is skipped for dynamic widgets already with the "continue" L2313.

Reply:
@ujfalusi https://github.com/thesofproject/sof/issues/6475 is still open. We know stress test fail with SOF2.0 and SOF2.1 and they don't fail with SOF2.2, but due to low repro rate, we've not been able to identify which commit fixes the problem (or if there are multiple).

Reply:
> @plbossart This is not about ABI, so the old/existing usage is a bit of a misusage of the SOF_ABI_VER macro. Old firmware version did implement the interfaces to free components, but they just didn't work. So problem is not in the interface, but rather in the implementation.
> 
> This is of course a bit academic as in practise referring to ABI always refers to a SOF version as well. But it does confuse the reader, so I'd rather fix this.

ok. I'd use clearer language than 'this is not linked to the ABI' in the commit message, somethink like 'In theory testing the ABI level should have been enough, in practice some versions of the firmware are broken so we use the firmware version here".